# Math 151B Homework No. 6

* Brandon Loptman  
* UID: 604105043  
* March 10, 2020  

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import scipy.linalg as la

### 9.3 The Power Method

#### 8.)

In [2]:
def PowerMethod(A,x,n,tol=10e-4, N=25):
    """
    Implements the Power Method as described by the algorithm in the textbook.
    """
    k = 1
    
    xp = np.max(abs(x))
    x = x/xp
    
    while(k <= N):
        #print(k)
        y = A.dot(x)
        
        yp = np.max(np.abs(y))
        u = yp
        #print(u)
        
        if(yp == 0):
            print("A has eigenvalue 0, select a new vector x and restart.")
            return x
        
        err = np.max(np.abs(x-(y/yp)))
        
        x = y/yp
        #print(x)
        
        if(err < tol):
            print("The procedure was successful!")
            return u,x
        
        k = k + 1
        
    print("The maximum number of iterations exceeded! The procedure was unsuccessful.")
    return u,x

#### (a.)

We want to find the dominant eignevalue and correspodning eigenvector of the matrix

\begin{pmatrix}
    4 & 2 & 1\\
    0 & 3 & 2\\
    1 & 1 & 4\\
\end{pmatrix}

using the Power Method.

In [3]:
A1 = np.array([[4,2,1],[0,3,2],[1,1,4]])
x1 = np.array([1,2,1])
n1 = 3

u,v = PowerMethod(A1,x1,n1)

print("Eigenvalue: ", u)
print("Eigenvetor: ", v)

The procedure was successful!
Eigenvalue:  5.915862109849257
Eigenvetor:  [1.         0.55415425 0.80959966]


We find that the dominant eigenvalue is $\lambda \approx 5.91586$ which corresponds to the eigenvector $\vec{v} \approx (1, 0.55415, .80960)^{T}$

#### (b.)

In [4]:
A2 = np.array([[1,1,0,0],[1,2,0,1],[0,0,3,3],[0,1,3,2]])
x2 = np.array([1,1,0,1])
n2 = 4

u,v = PowerMethod(A2,x2,n2)

print("Eigenvalue: ", u)
print("Eigenvetor: ", v)

The procedure was successful!
Eigenvalue:  5.666647958421926
Eigenvetor:  [0.0554278  0.25788625 1.         0.88873413]


#### (c.)

In [5]:
A3 = np.array([[5,-2,-0.5,1.5],[-2,5,1.5,-0.5],[-0.5,1.5,5,-2],[1.5,-0.5,-2,5]])
x3 = np.array([1,1,0,-3])
n3 = 4

u,v = PowerMethod(A3,x3,n3)

print("Eigenvalue: ", u)
print("Eigenvetor: ", v)

The procedure was successful!
Eigenvalue:  8.996446247045956
Eigenvetor:  [-0.99903998  0.9990224   0.9999824  -1.        ]


#### (d.)

In [6]:
A4 = np.array([[-4,0,0.5,0.5],[0.5,-2,0,0.5],[0.5,0.5,0,0],[0,1,1,4]])
x4 = np.array([0,0,0,1])
n4 = 4

u,v = PowerMethod(A4,x4,n4)

print("Eigenvalue: ", u)
print("Eigenvetor: ", v)

The maximum number of iterations exceeded! The procedure was unsuccessful.
Eigenvalue:  4.1202713104915745
Eigenvetor:  [0.10396255 0.07624767 0.01441345 1.        ]


#### 10.)

In [7]:
def InversePowerMethod(A,x,n,tol=10e-4,N=25):
    """
    Implements the Inverse Power Method as described by the algorithm in the textbook.
    """
    q = (np.transpose(x)).dot(A.dot(x))/(np.transpose(x)).dot(x)
    #print("q = ", q)
    
    k = 1
    
    xp = np.max(abs(x))
    x = x/xp
    
    qI = np.identity(n)*q
    
    while(k <= N):
        #print(k)
        
        y = np.linalg.solve(A-qI,x)
        #print(y)
        
        p = np.argmax(np.abs(y))
        yp = y[p]
        
        u = yp
        #print("u = ", u)
        
        err = np.max(np.abs(x-(y/yp)))
        x = y/yp
        #print(x)
        
        if(err < tol):
            u = (1/yp) + q
            #print("u = ", u)
            
            print("The procedure was successful!")
            return u,x
        
        k = k + 1
        
    print("The maximum number of iterations exceeded! The procedure was unsuccessful.")
    return u,x

#### (a.)

Now, we compute the Inverse Power Method to compute the dominant eigenvalues and corresponding eigenvectors for the same matricies from the previous problem.

In [8]:
u,v = InversePowerMethod(A1,x1,n1)

print("Eigenvalue: ", u)
print("Eigenvetor: ", v)

The procedure was successful!
Eigenvalue:  5.91952739334497
Eigenvetor:  [1.         0.55480496 0.80991748]


#### (b.)

In [9]:
u,v = InversePowerMethod(A2,x2,n2)

print("Eigenvalue: ", u)
print("Eigenvetor: ", v)

The procedure was successful!
Eigenvalue:  2.6459223598633343
Eigenvetor:  [ 0.60757625  1.         -0.32509997  0.03834611]


#### (c.)

In [10]:
u,v = InversePowerMethod(A3,x3,n3)

print("Eigenvalue: ", u)
print("Eigenvetor: ", v)

The procedure was successful!
Eigenvalue:  3.999973870192454
Eigenvetor:  [0.99999843 0.99993845 1.         0.99993999]


#### (d.)

In [11]:
u,v = InversePowerMethod(A4,x4,n4)

print("Eigenvalue: ", u)
print("Eigenvetor: ", v)

The procedure was successful!
Eigenvalue:  4.105293014436679
Eigenvetor:  [0.06281419 0.08704089 0.01825213 1.        ]


### 9.4 Householder's Method

#### 1.)

In [12]:
def HouseholderMethod(A,n):
    """
    Implements Householder's Method as described by the algorithm in the textbook.
    """

    for k in range(0,n-1):
        #print("k = ", k)
        
        u = np.zeros(n)
        v = np.zeros(n)
        z = np.zeros(n)
        q = 0
        
        for j in range(k+1,n):
            #print(A[j,k])
            q = q + (A[j,k]**2)
        
        #print("q = ", q)
        #print("A[k+1,k] = ",A[k+1,k])
        if(np.abs(A[k+1,k]) <= 10e-20):
            alpha = -1*np.sqrt(q)
        else:
            alpha = -1*(np.sqrt(q)*A[k+1,k]/np.abs(A[k+1,k]))
        
        #print("alpha = ",alpha)
        
        RSQ = (alpha**2) - alpha*A[k+1,k]
        if(RSQ == 0):
            RSQ = 10e-10
        #print("RSQ = ", RSQ)
        
        v[k] = 0
        v[k+1] = A[k+1,k] - alpha
        #print(A[k+1,k])
        
        #print("v = ", v)
        
        for j in range(k+2,n):
            #print("j = ", j)
            v[j] = A[j,k]
            
        #print("v = ", v)

        for j in range(k,n):
            #print("j = ", j)
            #print("--------------")
            for i in range(k+1,n):
                #print("i = ", i)
                #print("A[j,i] = ", A[j,i])
                #print("v[i] = ", v[i])
                u[j] = u[j] + A[j,i]*v[i]
            #print("--------------")
            #u[j] = u[j]/RSQ
        
        u = u/RSQ
        #print("u = ", u)
        
        PROD = 0
        for i in range(k+1,n):
            #print("i = ", i)
            PROD = PROD + (v[i]*u[i])
            
        #print("PROD = ", PROD)
        
        for j in range(k,n):
            #print("j = ", j)
            z[j] = u[j] - ((PROD/(2*RSQ))*v[j])
        #print("z = ", z)
            
        for l in range(k+1,n-1):
            #print("l = ", l)
            for j in range(l+1,n):
                #print("j = ", j)
                A[j,l] = A[j,l] - v[l]*z[j] - v[j]*z[l]
                A[l,j] = A[j,l]
                
            A[l,l] = A[l,l] - 2*v[l]*z[l]
            
        A[n-1,n-1] = A[n-1,n-1] - 2*v[n-1]*z[n-1]
        
        for j in range(k+2,n):
            #print("j = ", j)
            A[k,j] = 0
            A[j,k] = 0
            
        A[k+1,k] = A[k+1,k] - (v[k+1]*z[k])
        A[k,k+1] = A[k+1,k]
        
        #print(A)
                
    return A

#### (a.)

In [13]:
A1 = np.array([[12,10,4],[10,8,-5],[4,-5,3]])
n = 3

D = HouseholderMethod(A1,n)
print(D)

[[ 12 -10   0]
 [-10   3  -5]
 [  0  -5   7]]


For some reason this is missing the decimal portion of the matrix elements. The interger parts match up with the solution given in the textbook.

Both (b.) and (c.) do not match well with the solution given in the textbook. There is some issue with dividing by zero, but I'm not sure if it's an issue with my code or the problem is in the algorithm as stated in the textbook. Even when comparing my code with the MATLAB code on the publisher's website I can't find the error.

#### (b.)

In [14]:
A2 = np.array([[2,-1,-1],[-1,2,-1],[-1,-1,2]])

D = HouseholderMethod(A2,n)
print(D)

[[2 1 0]
 [1 0 0]
 [0 0 2]]


#### (c.)

In [15]:
A3 = np.array([[1,1,1],[1,1,0],[1,0,1]])

D = HouseholderMethod(A3,n)
print(D)

[[ 1 -1  0]
 [-1  0  0]
 [ 0  0  0]]


#### (d.)

In [16]:
A4 = np.array([[4.75,2.25,-0.25],[2.25,4.75,1.25],[-0.25,1.25,4.75]])

D = HouseholderMethod(A4,n)
print(D)

[[ 4.75       -2.26384628  0.        ]
 [-2.26384628  4.47560976  1.2195122 ]
 [ 0.          1.2195122   5.02439024]]


This is matches up well with the solution in the textbook.